src/parser.ts

```diff
@@ -306,17 +308,59 @@ function parseSuperTypes(
   options: Options,
   processed: Processed,
   usedNames: UsedNames
-): TNamedInterface[] {
+): TSuperType[] {
   // Type assertion needed because of dereferencing step
   // TODO: Type it upstream
   const superTypes = schema.extends as SchemaSchema | SchemaSchema[] | undefined
   if (!superTypes) {
     return []
   }
   if (Array.isArray(superTypes)) {
-    return superTypes.map(_ => newNamedInterface(_, options, _, processed, usedNames))
+    return superTypes.map(_ => getOrCreateSuperType(_, options, _, processed, usedNames))
+  }
+  return [getOrCreateSuperType(superTypes, options, superTypes, processed, usedNames)]
+}
+
+function isSuperType(type: any): type is TSuperType {
+  return type.type === 'CUSTOM_TYPE' || type.type === 'INTERFACE'
+}
+
+function getOrCreateSuperType(
+  schema: SchemaSchema,
+  options: Options,
+  rootSchema: JSONSchema,
+  processed: Processed,
+  usedNames: UsedNames
+): TSuperType {
+  if (processed.has(schema)) {
+    const hit = processed.get(schema)
+    if (isSuperType(hit)) {
+      return hit
+    }
+  }
+
+  if (schema.tsType) {
+    const extendedType = newExtendedType(schema, usedNames)
+    processed.set(schema, extendedType)
+    return extendedType
+  }
+  const namedInterface = newNamedInterface(schema, options, rootSchema, processed, usedNames)
+  processed.set(schema, namedInterface)
+  return namedInterface
+}
+
+function newExtendedType(schema: SchemaSchema, usedNames: UsedNames): TCustomType {
```

Comment:
Can we tighten up the types here?

Then, you can remove the if/else/throw below.

```suggestion
function newExtendedType(schema: CustomTypeJSONSchema, usedNames: UsedNames): TCustomType {
```

Reply:
Alright